x/exchange/keeper/grpc_query_test.go

```diff
@@ -1,27 +1,4017 @@
+					FeeBuyerSettlementRatios:  s.ratios("1000pineapple:3pineapple,100prune:3prune"),
+					AcceptingOrders:           true,
+					AllowUserSettlement:       true,
+					AccessGrants: []exchange.AccessGrant{
+						{Address: s.addr1.String(), Permissions: exchange.AllPermissions()},
+						{Address: s.addr2.String(), Permissions: []exchange.Permission{1, 2}},
+						{Address: s.addr3.String(), Permissions: []exchange.Permission{3, 4}},
+						{Address: s.addr4.String(), Permissions: []exchange.Permission{5, 6}},
+						{Address: s.addr5.String(), Permissions: []exchange.Permission{2, 4, 6, 7}},
+					},
+					ReqAttrCreateAsk: []string{"ask.good.kyc", "*.my.custom"},
+					ReqAttrCreateBid: []string{"bid.good.kyc", "*.my.custom"},
+				})
+				s.requireCreateMarketUnmocked(exchange.Market{MarketId: 4})
+				s.requireCreateMarketUnmocked(exchange.Market{MarketId: 5})
+			},
+			req: &exchange.QueryGetMarketRequest{MarketId: 3},
+			expResp: &exchange.QueryGetMarketResponse{
+				Address: exchange.GetMarketAddress(3).String(),
+				Market: &exchange.Market{
+					MarketId: 3,
+					MarketDetails: exchange.MarketDetails{
+						Name:        "Market Three",
+						Description: "This is the third market. Not the first or second. And fourth is just too far.",
+						WebsiteUrl:  "not actually a websute url for market 3",
+						IconUri:     "https://www.example.com/market/3/icon",
+					},
+					FeeCreateAskFlat:          s.coins("10fig,100grape"),
+					FeeCreateBidFlat:          s.coins("20fig,200grape"),
+					FeeSellerSettlementFlat:   s.coins("10pineapple,50prune"),
+					FeeSellerSettlementRatios: s.ratios("1000pineapple:1pineapple,100prune:1prune"),
+					FeeBuyerSettlementFlat:    s.coins("12pineapple60prune"),
+					FeeBuyerSettlementRatios:  s.ratios("1000pineapple:3pineapple,100prune:3prune"),
+					AcceptingOrders:           true,
+					AllowUserSettlement:       true,
+					AccessGrants: []exchange.AccessGrant{
+						{Address: s.addr1.String(), Permissions: exchange.AllPermissions()},
+						{Address: s.addr2.String(), Permissions: []exchange.Permission{1, 2}},
+						{Address: s.addr3.String(), Permissions: []exchange.Permission{3, 4}},
+						{Address: s.addr4.String(), Permissions: []exchange.Permission{5, 6}},
+						{Address: s.addr5.String(), Permissions: []exchange.Permission{2, 4, 6, 7}},
+					},
+					ReqAttrCreateAsk: []string{"ask.good.kyc", "*.my.custom"},
+					ReqAttrCreateBid: []string{"bid.good.kyc", "*.my.custom"},
+				},
+			},
+		},
+	}
+
+	for _, tc := range tests {
+		s.Run(tc.name, func() {
+			runQueryTestCase(s, testDef, tc)
+		})
+	}
+}
+
+func (s *TestSuite) TestQueryServer_GetAllMarkets() {
+	briefIDStringer := func(brief *exchange.MarketBrief) string {
+		if brief == nil {
+			return "<nil>"
+		}
+		return fmt.Sprintf("%d", brief.MarketId)
+	}
+	testDef := queryTestDef[exchange.QueryGetAllMarketsRequest, exchange.QueryGetAllMarketsResponse]{
+		queryName: "GetAllMarkets",
```

Comment:
I doubt this has the same issue as orders, but will this always be usable without pagination?

Reply:
This one also has pagination on it already. It'll probably be a long time until there's enough markets for it to be unusable without pagination, but any time there's a variable number of results in a query, it's best to include pagination.